hibernate-core/src/main/java/org/hibernate/loader/Loader.java

```diff
@@ -1215,8 +1242,13 @@ private void initializeEntitiesAndCollections(
 			}
 		}
 
-		if ( hydratedObjects != null ) {
-			for ( Object hydratedObject : hydratedObjects ) {
+		if ( nullSeparatedHydratedObjects != null ) {
+			for ( Object hydratedObject : nullSeparatedHydratedObjects ) {
+				if ( hydratedObject == null ) {
+					// This is a hack to signal that we're starting to process a new row
+					// Ignore
+					continue;
+				}
```

Comment:
Did you conduct some MPH benchmark comparison to justify the hack? It would be great to avoid such hack for the entity graph implementation logic has been so difficult to understand and maintain.

Reply:
I did not. I'm just considering that keeping the same data structure (a List) and adding a few elements is less change than moving to a `List<List<?>>`. I'm not saying it's the same, mind you. Just that it's the closest I can get.

Reply:
I am not against such hack as long as it turns out to be the hot spot. Also, the fetch graph impl in v5 has been based on hack already. Overall the PR looks good from a quick review. I might focus on it more later in my off hours to finalize the review. Thanks for the bug reporting and quick fixing.